File: Add Code Here/C/wordbreak.c

class Solution{
public:
    vector<string> ans;
    void check(string s,int n,unordered_set<string> u,string tmp)
    {
        if(s.length()==0)
        {
            tmp.pop_back();
            ans.push_back(tmp);
            return;
        }
        for(int i=1;i<=s.length();i++)
        {
            string l=s.substr(0,i);
            string r=s.substr(i);
            if(u.find(l)!=u.end())
            check(r,n,u,tmp+l+" ");
        }
    }
    vector<string> wordBreak(int n, vector<string>& dict, string s)
    {
        //can use recursive method 
        unordered_set<string> u;
        for(int i=0;i<n;i++)
        u.insert(dict[i]);
        check(s,n,u,"");
        return ans;
    }
};

File: Add Code Here/JAVA/Word Break.java
public class Solution {
    public int wordBreak(String A, String[] B) {
        Set<String> set = new HashSet<>();
        for (String s : B) {
            set.add(s);
        }
        Boolean[] dp = new Boolean[A.length()];
        return wordBreak(A, set, 0, dp);
    }
    
    private int wordBreak(String s, Set<String> set, int start, Boolean[] dp) {
        if (start == s.length()) {
            return 1;
        }
        if (dp[start] != null) return dp[start] ? 1 : 0;
        
        int ans = 0;
        for (int end = start + 1; end <= s.length(); end++) {
            if (set.contains(s.substring(start, end)) && wordBreak(s, set, end, dp) == 1) {
                ans = 1;
                break;
            }
        }
        
        dp[start] = ans == 1;
        return ans;
    }
}

